### PR TITLE
fix(ai): close 'prompt is too long' bug class structurally

### PR DESCRIPTION
**Pull request by kylejryan**
## Summary

Production was emitting recurring context-length errors:

```
[AI_APICallError]   prompt is too long: 201755 tokens > 200000 maximum
[AI_RetryError]     Failed after 2 attempts ... prompt is too long: 210190 tokens > 200000 maximum
[AI_RetryError]     Failed after 2 attempts ... prompt is too long: 213021 tokens > 200000 maximum
[AI_ToolCallRepairError] Error repairing tool call: ...
```

The 1–7% overshoots (201K → 213K against a 200K window) traced to three compounding causes. This PR replaces the heuristics with structural invariants so the bug class can't recur — every code path that can construct provider input now either fits-by-construction or escalates with a typed terminal signal.

## Root cause analysis

The previous fix layered Layer 1 (tool-result truncation) + Layer 2 (step snipping) + Layer 3 (full summarization), and they did work — when invoked. But three things let oversized prompts still reach the provider:

### 1. Tool-schema overhead was not counted toward the budget
`fitMessagesToContext` reserved a fixed `5_000`-token constant for "everything else" (tool definitions + envelope). Apex's offensive-security agent has 15+ tools with detailed descriptions and Zod schemas — actual schema overhead is **15–25K tokens**, not 5K. So the proactive fit thought it had ~120K tokens of headroom for messages when reality was ~100K. A history that fit "comfortably" by our math overflowed the provider by exactly the gap (~1–7%, matching the production logs).

### 2. Reactive-only compaction (`fitsBudget=false` was treated as advisory)
When `fitMessagesToContext` returned `fitsBudget: false`, `streamResponse` still passed those messages to `streamText`. The doomed call hit the provider, the SDK threw `AI_APICallError`, and only **then** did recovery run. Every overshoot in the logs was a "send-and-pray" pattern at one level deeper than the user-visible call site.

### 3. Recovery paths could themselves overflow
- `summarizeConversation` embedded the full `opts.system` (often 50K+ chars for the operator/pentest prompt) verbatim in a user message, then sliced the last 20 messages with no size cap. A single 100K-char assistant turn walked through.
- `experimental_repairToolCall` embedded `toolCall.input` raw — a tool that takes a multi-megabyte content field (`write_file`, `http_request` body, `execute_command` heredocs) blew the repair `generateText` call.
- The `summarize → resume → still over budget → summarize` chain had no termination condition; it could loop indefinitely or stack-overflow.

## How this fixes the issue

The fix converts "context fits the budget" from a heuristic property to a structural invariant. Every site that can send oversized input is guarded by a check that either fits it or escalates with a typed signal — never sends and prays.

### Invariants now enforced

| # | Invariant | Enforcement |
|---|-----------|-------------|
| **I1** | `estimate(input) + max_output + safety ≤ contextWindow` before any provider call | `streamResponse` calls `fitMessagesToContext` proactively; `fitsBudget=false` routes to summarization **before** `streamText` (closes RCA 2) |
| **I2** | Compaction is monotone — never claims success on unchanged-and-still-over input | `fitMessagesToContext` returns `{ messages, modified, fitsBudget }`; tested |
| **I3** | Auxiliary LLM calls (summarization, repair) have hard caps independent of caller input | `MAX_PER_MESSAGE_CHARS = 4_000`, `MAX_SLICED_TOTAL_CHARS = 25_000`, `SYSTEM_PROMPT_SNIPPET_CHARS = 2_000`, `MAX_TOOL_INPUT_CHARS = 8_000`, `MAX_SCHEMA_CHARS = 6_000` (closes RCA 3, parts a/b) |
| **I4** | Any failure in any context-recovery path is caught at the `streamResponse` boundary | `wrapStreamWithErrorHandler` wraps **both** the post-`streamText` path AND the proactive-summarization escalation. Pinned by `ai.proactive-recovery.test.ts`. |
| **I5** | `streamResponse` ↔ Pensar gateway formatter agree on `max_output_tokens` | Single source of truth: `models/index.ts:getMaxOutputTokens`, imported by both consumers. Self-clamps against `contextLength` (catches legacy Bedrock models with tiny windows). |
| **I6** | Every model in `AVAILABLE_MODELS` has a sane budget | Exhaustive registry test (`models.test.ts`) iterates registry; fails if any model has `max ≥ ctx` or insufficient input headroom. |
| **I7** | `streamResponse` always counts tool-schema overhead toward the budget | `fitMessagesToContext` accepts `tools: ToolSet`; `estimateToolsOverheadTokens` derives overhead from the actual schemas (closes RCA 1). |
| **I8** | Recursive recovery is bounded | `_restartDepth` plumbed through `StreamResponseOpts`, bumped at every summarize→resume boundary. Past `MAX_RESTART_DEPTH=3`, throws typed `ContextLengthExhaustedError`. |

### State machine

```
[pre-flight fit] ──fits──→ send ─ok─→ done
       │                     │ ctx_err
       │                     ↓
       │              [reactive fit]
       │                fits → retry → done
       │                not_fits → Layer 3 summarize
       │                              ok → continue
       │                              fail → minimal-restart (bumps _restartDepth)
       │                                       depth ≤ MAX → loop
       │                                       depth > MAX → ContextLengthExhaustedError (typed terminal)
       └──not_fits──→ proactive summarize (also wrapped) → same recovery as above
```

Every edge that touches a provider goes through `fitMessagesToContext` or a hard char cap. The only context-related error allowed to escape `streamResponse` is the typed `ContextLengthExhaustedError` — never a raw `AI_APICallError prompt is too long`.

## Files changed

| File | Net change | What it does |
|------|------------|---------------|
| `src/core/ai/contextManagement.ts` | +258 | New `fitMessagesToContext` primitive; token estimators including `estimateToolsOverheadTokens` (memoized via `WeakMap<ToolSet, number>`); shared `truncateWithMarker` helper; Layer 1 cascading thresholds share `persistedIds: Set` so the same toolCallId is written to disk at most once per fit |
| `src/core/ai/ai.ts` | +193 / −0 (net) | Proactive escalation; reactive recovery uses same primitive; bounded repair-call inputs; `_restartDepth` plumbing; typed `ContextLengthExhaustedError`; proactive-summarization stream wrapped through `wrapStreamWithErrorHandler` |
| `src/core/ai/utils.ts` | +65 / −0 (net) | Bounded `summarizeConversation` (per-message + cumulative caps + system snippet); depth bump on resumed call |
| `src/core/ai/models/index.ts` | +50 | `getMaxOutputTokens` single-source-of-truth + self-clamp |
| `src/core/ai/providers/pensarFormatters.ts` | +35 / −35 | Removes duplicate `getDefaultMaxOutputTokens`; imports canonical helper |

### New tests

| File | Pins |
|------|------|
| `src/core/ai/contextManagement.test.ts` | The primitive: budget math, Layer 1 triggers on huge tool results, tools-overhead derived from `ToolSet`, reference-equality on no-op |
| `src/core/ai/ai.contract.test.ts` | Recursion bound: `_restartDepth > MAX` → synchronous `ContextLengthExhaustedError` carrying the depth |
| `src/core/ai/ai.proactive-recovery.test.ts` | I4 on the proactive path: a context error inside the proactive-summarization stream is caught by the wrapper; exhausted recovery surfaces `ContextLengthExhaustedError`, never the raw provider error |
| `src/core/ai/models/models.test.ts` | Every model has `max < ctx` and ≥50K input headroom; tier-specific Claude budgets are pinned |

## Test plan

- [x] `bun run lint` — clean
- [x] `bun run format:check` — clean
- [x] `bun run test` — 959 passed, 16 skipped (pre-existing API-key-gated integration tests)
- [ ] Production validation after merge:
  - Day-1 log search for `AI_APICallError.*prompt is too long` should drop to zero
  - Watch for any new `ContextLengthExhaustedError` (typed terminal — signals the agent hit a genuinely unrecoverable input shape, surfaces a clean error rather than a raw SDK trace)
  - Watch the `Proactive context fit returned fitsBudget=false` warning as an early signal of input shapes that needed escalation

## Risks

| Risk | Mitigation | Severity |
|------|-----------|----------|
| chars/4 token estimate drifts on code-dense content | 10K-token safety margin + reactive path with same estimator + depth bound | Low |
| Proactive escalation fires on borderline cases that previously slipped through | Those would have failed anyway; this prevents the bad call instead of recovering from it | None — strictly an improvement |
| `MAX_RESTART_DEPTH=3` caps recovery in pathological cases | Typed error preferable to infinite loop; 3 cycles is generous | Low |
| `getMaxOutputTokens` clamp tightens output on legacy Bedrock models (e.g. cohere with ctx=4000) | Not actively used in Apex routing | None |

## Out of scope (follow-up)

The structural fix closes the production bug class. The following are quality-of-life follow-ups, not blockers:

- **Layer 2.5**: per-tool-call-input + per-text-part caps in conversation history (closes the theoretical "single huge agent-generated message" path; would convert any remaining `ContextLengthExhaustedError` from agent self-growth into transparent recovery)
- **Per-tool structure-aware Layer 1 previews**: replace head-only truncation with extractor-per-tool (status/exit-code/findings preservation)
- **`read_tool_result` tool**: lets the agent fetch the full disk-persisted output of a truncated tool call by `toolCallId` when it realizes truncation removed something it needs

Made with [Cursor](https://cursor.com)

---

> [!NOTE]
> **Medium Risk**
> Touches core `streamResponse` request/recovery logic and model output-token budgeting, which can change production behavior (when/if summarization triggers and how errors propagate). Main risk is unintended escalation/termination in edge cases or overly aggressive truncation, though changes are heavily pinned by new unit tests.
> 
> **Overview**
> Prevents oversized prompts from reaching providers by adding **proactive context fitting** before `streamText` and escalating directly to summarization when Layer 1+2 can’t fit the budget.
> 
> Hardens the context-length recovery state machine: all summarization entry points (proactive, reactive, and outer `try/catch`) are now wrapped so raw provider “prompt is too long” errors don’t leak, recovery recursion is **bounded** via `_restartDepth` and a typed terminal `ContextLengthExhaustedError`, and summarization failures fall back to a minimal-context restart.
> 
> Improves budget correctness and safety caps: introduces `fitMessagesToContext` with token estimation that includes **tool/schema overhead**, makes Layer 1 truncation cascaded/monotone with correct metadata + deduped persistence, adds strict char caps for summarization and tool-call repair inputs, and centralizes per-model `maxOutputTokens` via `getMaxOutputTokens` (used by both `streamResponse` and Pensar formatting) with exhaustive tests pinning model budgets and invariants.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 73586546f7b5eb68741847feef283a2279cbfdad. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/core/ai/contextManagement.ts [hunk 1]
```diff
@@ -84,30 +109,26 @@ export function applyToolResultBudget(
...
-const DEFAULT_KEEP_RECENT_STEPS = 15;
+// Larger windows can blow the budget on the "recent" slice alone; 6 leaves
+// enough recent context to make progress while letting Layer 2 actually compact.
+const DEFAULT_KEEP_RECENT_STEPS = 6;
 
 /**
  * For tool results older than `keepRecentSteps`, replace with 1-line summaries.
- * Keeps full content for recent steps to preserve context for active work.
- *
- * Returns a new messages array (does not mutate the input).
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/contextManagement.ts:102]: @cursoragent Fix

**Review by joshkotrous — APPROVED** [review]

File: src/core/ai/contextManagement.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshkotrous) [reply src/core/ai/contextManagement.ts:102]: @cursor I SAID FIX

File: src/core/ai/ai.ts [hunk 2]
```diff
@@ -695,9 +791,9 @@ export function streamResponse(
               prompt: [
                 `The model tried to call the tool "${toolCall.toolName}"` +
                   ` with the following inputs:`,
-                toolCall.input,
+                boundedInput,
                 `The tool accepts the following schema:`,
-                JSON.stringify(jsonSchema),
+                boundedSchema,
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/ai.ts:865]: @cursoragent Fix thsi

File: src/core/ai/models/index.ts [hunk 3]
```diff
@@ -31,3 +31,53 @@ export function getModelInfo(model: AIModel): ModelInfo {
...
+    return 64_000;
+  }
+  if (modelId.includes("claude-opus-4-")) {
+    return 32_000;
+  }
+  if (modelId.includes("claude-3-5-haiku")) {
+    return 8_192;
+  }
+  return 4_096;
+}
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/models/index.ts:159]: @cursoragent Fix

File: src/core/ai/ai.ts [hunk 4]
```diff
@@ -354,35 +360,29 @@ function wrapStreamWithErrorHandler(
...
                     console.warn(
-                      `Context length error — trying lightweight compaction on ${currentMessages.length} messages`,
+                      `Context length error — Layer 1+2 reduced to ~${fitted.estimatedInputTokens} tokens, retrying`,
                     );
                   }
                   try {
                     const retried = streamResponse({
                       ...opts,
-                      messages: compacted,
+                      messages: fitted.messages,
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/ai.ts:402]: @cursoragent Fix

File: src/core/ai/contextManagement.ts [hunk 5]
```diff
@@ -1,40 +1,64 @@
...
+  text: string,
+  max: number,
+  label?: string,
+): string {
+  if (text.length <= max) return text;
+  const marker = label
+    ? `…[${label} truncated]`
+    : `…[truncated ${text.length - max} chars]`;
+  return `${text.slice(0, max)}${marker}`;
+}
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/contextManagement.ts:48]: @cursoragent Fix

File: src/core/ai/contextManagement.ts [hunk 6]
```diff
@@ -46,34 +70,40 @@ export function applyToolResultBudget(
...
+      // making the "truncated" message larger than the original and
+      // reporting a negative dropped-char count.
+      const previewSize = Math.min(maxChars, 2000);
+      const preview = text.slice(0, previewSize);
       return {
         ...p,
         output: {
           type: "text" as const,
-          value: `${preview}\n\n[... truncated ${text.length - 2000} chars — full output saved to ${filePath}]`,
+          value: `${preview}\n\n[... truncated ${text.length - previewSize} chars — full output saved to ${filePath}]`,
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/contextManagement.ts]: @cursoragent Fix

File: src/core/ai/ai.ts [hunk 7]
```diff
@@ -398,29 +407,56 @@ function wrapStreamWithErrorHandler(
...
-                  yield chunk;
+                try {
+                  const summarizationStream = createSummarizationStream(
+                    messagesForSummary,
+                    opts,
+                    model,
+                  );
+                  for await (const chunk of summarizationStream.fullStream) {
+                    yield chunk;
+                  }
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/ai.ts:464]: @cursoragent Fix

File: src/core/ai/models/index.ts [hunk 8]
```diff
@@ -31,3 +31,59 @@ export function getModelInfo(model: AIModel): ModelInfo {
...
+  ) {
+    return 64_000;
+  }
+  if (modelId.includes("claude-opus-4-")) {
+    return 32_000;
+  }
+  if (modelId.includes("claude-3-5-haiku")) {
+    return 8_192;
+  }
+  return 4_096;
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/models/index.ts:158]: @cursoragent Fix

File: src/core/ai/ai.ts [hunk 9]
```diff
@@ -546,18 +609,60 @@ export function streamResponse(
...
+  // an overflow inside the summarization call itself escapes uncaught while
+  // the equivalent post-streamText path (line ~880) recovers.
+  if (proactiveFitFailed && fittedMessages) {
+    if (!silent) {
+      console.warn(
+        `Proactive context fit returned fitsBudget=false on ${fittedMessages.length} messages — escalating to summarization before send`,
+      );
+    }
+    return wrapStreamWithErrorHandler(
+      createSummarizationStream(fittedMessages, opts, providerModel),
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/ai.ts:692]: @cursoragent Fix

**Comment by kylejryan** [comment]
@cursoragent Babysit this PR and resolve the bug bot issues until the build is fully green

File: src/core/ai/ai.ts [hunk 10]
```diff
@@ -784,11 +918,21 @@ export function streamResponse(
...
+      return wrapStreamWithErrorHandler(
+        createSummarizationStream(
+          messagesContainer.current,
+          opts,
+          providerModel,
+        ),
+        messagesContainer,
         opts,
         providerModel,
+        silent,
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/ai.ts:971]: @cursoragent Fix

File: src/core/ai/models/index.ts [hunk 11]
```diff
@@ -31,3 +31,117 @@ export function getModelInfo(model: AIModel): ModelInfo {
...
+    modelId.includes("gemini-2.0") ||
+    modelId.includes("gemini-1.5") ||
+    modelId.includes("gemini-pro") ||
+    modelId.includes("gemini-flash")
+  ) {
+    return 8_192;
+  }
+
+  return 4_096;
+}
```
[the remark this answers is not in the document]
Comment (kylejryan) [reply src/core/ai/models/index.ts:159]: @cursoragent Fix
